File: Trees/104. Maximum Depth of Binary Tree/Solution.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # N : no of nodes in the tree
        # Time Complexity: O(N)
        # Space Complexity: (N) -> Stack Space
        # def dfs(root):
        #     if not root:
        #         return 0
            
        #     left = dfs(root.left)
        #     right = dfs(root.right)

        #     return max(left, right) + 1
        # return dfs(root)


        # N : no of nodes in the tree
        # Time Complexity: O(N)
        # Space Complexity: (N)
        # if not root:
        #     return 0
        
        # stack = deque()
        # stack.append([root, 1])
        # depth = 0
        # while stack:
        #     node, height = stack.pop()
        #     depth = max(depth, height)
        #     if node.right:
        #         stack.append([node.right, height + 1])
        #     if node.left:
        #         stack.append([node.left, height + 1])
        # return depth



        # N : no of nodes in the tree
        # Time Complexity: O(N)
        # Space Complexity: (N)
        if not root:
            return 0

        q = deque([root])
        depth = 0
        while q:
            depth += 1
            n = len(q)
            for _ in range(n):
                node = q.popleft()

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return depth

File: Trees/104. Maximum Depth of Binary Tree/Solution.java
import java.util.*;
import java.lang.Math;

// Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class Solution {
    private int dfs(TreeNode root) {
        if (root == null) return 0;

        int left = dfs(root.left);
        int right = dfs(root.right);

        return Math.max(left, right) + 1;
    }

    public int maxDepth(TreeNode root) {
        // N : no of nodes in the tree
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        // return dfs(root);


        // N : no of nodes in the tree
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        // if (root == null) return 0;

        // Deque<Pair<TreeNode, Integer>> stack = new ArrayDeque<>();
        // stack.push(new Pair<>(root, 1));
        
        // int depth = 0;
        // while (!stack.isEmpty()) {
        //     Pair<TreeNode, Integer> pair = stack.pop();
        //     depth = Math.max(depth, pair.getValue());
            
        //     if (pair.getKey().right != null) stack.push(new Pair<>(pair.getKey().right, pair.getValue() + 1));
        //     if (pair.getKey().left != null) stack.push(new Pair<>(pair.getKey().left, pair.getValue() + 1));
        // }
        // return depth;


        // N : no of nodes in the tree
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        if (root == null) return 0;

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        int depth = 0;
        while (!q.isEmpty()) {
            depth++;
            
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode node = q.poll();
                if (node.left != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
        }
        return depth;
    }
}